### ISSUE TITLE
hir::Type::impls_trait false positive

### ISSUE THREAD

**Issue by A4-Tacks** [rust-lang/rust-analyzer#20574]
**rust-analyzer version**: rust-analyzer 0.4.2595-standalone

**rustc version**: rustc 1.91.0-nightly (040a98af7 2025-08-20)

**editor or extension**: VIM9.1

**code snippet to reproduce**:
```rust
struct Foo<T>($0T);
```

**Output assists**:

```
╭Choose action─────────────────────╮
│1. Change visibility to pub(crate)│
│2. Add `#[derive]`                │
│3. Generate impl for `Foo`        │
│4. Generate trait impl for `Foo`  │
│5. Generate `new`                 │
╰──────────────────────────────────╯
```

`` Generate `Deref` not exist ``

**Output assists (rust-analyzer 1.91.0-nightly (040a98a 2025-08-20)**:

```
╭Choose action─────────────────────╮
│1. Change visibility to pub(crate)│
│2. Generate `Deref` impl using `T`│
│3. Add `#[derive]`                │
│4. Generate impl for `Foo`        │
│5. Generate trait impl for `Foo`  │
│6. Generate `new`                 │
╰──────────────────────────────────╯
```

This problem from `crates/ide-assists/src/handlers/generate_deref.rs ` line 172:

```rust
    if strukt_type.impls_trait(sema.db, deref_trait, &[]) {
        if strukt_type.impls_trait(sema.db, deref_mut_trait, &[]) {
            Some(DerefType::DerefMut)
        } else {
            Some(DerefType::Deref)
        }
    } else {
        None
    }
```

For has generic parameter type, `impls_trait` seems to have unexpectedly returned true

**Comment by ChayimFriedman2** [comment]
This depends on how exactly you retrieved `strukt_type`. If its generic args are the identity args, then this check will return false (unless the trait is implemented for any args). If it is instantiated with error types (which I suspect is what you had), it will return true, since the error types are converted into infer vars. Since the recent refactoring of `hir::Type`, the default choice is to leave the identity args (instantiation is explicit).

As an aside, if you want the behavior you observed, aside from instantiating with error types you can now also call `has_any_impl()` (without instantiating with errors), which is different in a subtle but important way: it allocates the same infer var for the same params. For example (a real example from an assist), if you want to check whether there exist `T` such that `Struct<T>: From<Struct<T>::FieldType>`, instantiating with errors will allow the two `T`s to be different (e.g., if `FieldType == T`, it will approve `Struct<i32>: From<bool>`), while `has_any_impl()` won't.
